src/compiler/checker.ts

```diff
@@ -5027,6 +5027,7 @@ namespace ts {
         function getTypeFromTupleTypeNode(node: TupleTypeNode): Type {
             const links = getNodeLinks(node);
             if (!links.resolvedType) {
+                checkTupleType(node);

```

Comment:
That's not the right fix. This function is simply for materializing the type, it isn't for the actual checking. The real problem is that `checkAssertion` never calls `checkSourceElement` on `node.type`. You can prove that to yourself by observing that this doesn't generate an error either:

```ts
var x = <{ a: number, a: number}>undefined;  // No duplicate identifier error
```

So, remove the line above and instead put a call to `checkSourceElement(node.type)` in `checkAssertion`.